### DISCUSSION TITLE
CF flipped once takeoff

### DISCUSSION THREAD

**Discussion by yubinwang11** [USC-ACTLab/crazyswarm#499] [General]
Hi @whoenig and @jpreiss,

I am facing trouble when testing one cf flight.
My implementation includes: 

1. optitrack mocap
2. my hover launch file as following
`
<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />

  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflieTypes.yaml" />
  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflies.yaml" />

  <node pkg="crazyswarm" type="crazyswarm_server" name="crazyswarm_server" output="screen" >
    <rosparam>
      world_frame: "/world"
      # Logging configuration (Use enable_logging to actually enable logging)
      genericLogTopics: ["log1"]
      genericLogTopicFrequencies: [10]
      genericLogTopic_log1_Variables: ["stateEstimate.x", "ctrltarget.x"]
      # firmware parameters for all drones (use crazyflieTypes.yaml to set per type, or
      # allCrazyflies.yaml to set per drone)
      firmwareParams:
        commander:
          enHighLevel: 1
        stabilizer:
          estimator: 2 # 1: complementary, 2: kalman
          controller: 2 # 1: PID, 2: mellinger
        ring:
          effect: 16 # 6: double spinner, 7: solid color, 16: packetRate
          solidBlue: 255 # if set to solid color
          solidGreen: 0 # if set to solid color
          solidRed: 0 # if set to solid color
          headlightEnable: 0
        locSrv:
          extPosStdDev: 1e-3
          extQuatStdDev: 0.5e-1
        kalman:
          resetEstimation: 1
      # tracking
      motion_capture_type: "optitrack" 
      motion_capture_host_name: "192.168.1.200" # hostname or IP address

      object_tracking_type: "motionCapture" 
      send_position_only: False # set to False to send position+orientation; set to True to send position only

      optitrack_local_ip: "192.168.1.102"
      optitrack_server_ip: "192.168.1.200"

      optitrack_host_name: "192.168.1.200"

      save_point_clouds: ~/pointCloud.ot
      print_latency: False
      write_csvs: False
      force_no_cache: False
      enable_parameters: True
      enable_logging: False

      #print_latency: True

      
    </rosparam>
  </node>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
    <param name="dev" value="$(arg joy_dev)" />
  </node>

  <node pkg="crazyswarm" type="crazyswarm_teleop" name="crazyswarm_teleop" output="screen">
    <param name="csv_file" value="$(find crazyswarm)/launch/figure8_smooth.csv" />
    <param name="timescale" value="0.8" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyswarm)/launch/test.rviz"/>

  
  

</launch>

`
3. unique marker configuration with 3 markers
4. the LED ring is solid green shows that the connection quality is good at least until it crashed
5. the up axis in motive is z up.
6. the direction of front cf points to x axis
7. alreadly flashed the latest firmware which is the version of June.2021

And the real flight is same as the video previously mentioned in another issue, previous video is [here](https://www.youtube.com/watch?v=dJ-tqMDjkeY)
Sorry to disturb u and I was stuck here for a while.

my guess:
1. optitrack cannot recognize the cf after taking off
2. the communication was cut off.
3. the oritention is false
But all things were considered and I kept avoiding this.

Looking forward to your suggestions.

**Comment by whoenig** [comment]
* You don't need `optitrack_local_ip` and `optitrack_local_ip`
* Since the green LED stays on, your communication is fine
* It seems like a problem with your orientation. Manually move your drone around (not flying; with your hands) and check the movement in rviz: if you move forward, x should increase; if you move left, y should increase; if you move up, z should increase. Front is where the antenna is and you have to consider a right-handed coordinate system.
* You may also try setting `send_position_only` to `True` to check if it's a problem with your orientation estimate.

**Comment by jpreiss** [comment]
Also double-check that your marker arrangement is not symmetric.

**Comment by yubinwang11** [comment]
Hi @whoenig @jpreiss  again,

Thanks for ur quick reply.
I figured out the reason, that's because the controller parameter is not optimal after replacing with larger baterry. 
But the trouble now is unstable flight after taking off especially during landing.
I plan to adjust parameter in your controller for the potential unstablity caused by mass. And already changed the mass and masstrust as following.
`crazyflieTypes:    
  default:      
    bigQuad: False    
    batteryVoltageWarning: 3.8  # V    
    batteryVoltateCritical: 3.7 # V     
    markerConfiguration: 0 # change to 3 for single marker tracking    
    dynamicsConfiguration: 0    
    firmwareParams:    
      ctrlMel:    
        kp_xy: 0.4 #0.6 # p for position    
        kd_xy: 0.2 # 0.35 # d for position    
        ki_xy: 0.05   
        i_range_xy: 2.0   
        kR_xy: 70000 # p for attitude   
        kw_xy: 20000 # d for attitude   
        kR_z: 60000 # 70000 # p for yaw   
        kw_z: 12000 # 30000  # d for yaw   
        ki_m_z: 500 #0 #20000     # i for yaw   
        i_range_m_z: 1500 # i range for yaw   
        kd_omega_rp: 200 # roll and pitch angular velocity d gain   
        kp_z: 1.25   
        kd_z: 0.4   
        ki_z: 0.05   
        i_range_z: 0.4   
        mass: 0.040   
        massThrust: 140000 #132000
`
But this is not enough for stable flight. Is it possible for modifying aggressive parameter? And any other suggestions?

**Reply by whoenig** [reply]
Try the PID controller. It's not that great at tracking, but more forgiving if the dynamics do not match. Other than that, you likely need to retune the complete controller with a significant mass increase. I am a bit surprised that 40g gives you that much problems. I certainly tested the default configuration up to 36g and it works ok, as long as you don't try to fly fast. Note that the thrust-to-weight ratio is only about 1.4 (measured with a mass of 33g), so you are very close to the limit of the standard motors. You could consider upgrading your motors to get better flight performance, too.

**Reply by yubinwang11** [reply]
To be more specific, if default controller and default parameters, 37g mass pushes me into trouble. This not fair at all. And PID also cannot ensure a stable flight. Maybe the battery level is another reason. Replacing more forceful motors seems a promising way. :)

**Reply by jpreiss** [reply]
Sounds like a battery problem. How old are your batteries? What is their rated discharge capacity (a 2-digit number followed by "c")? Old LiPo batteries will charge to the normal voltage, but then have a huge voltage drop under load.
